src/io/flutter/jxbrowser/JxBrowserManager.java

```diff
@@ -0,0 +1,173 @@
+    for (File file : files) {
+      if (file.exists()) {
+        if (!file.delete()) {
+          LOG.info(project.getName() + ": Existing file could not be deleted - " + file.getAbsolutePath());
+        }
+      }
+    }
+
+    downloadJxBrowser(project, fileNames);
+  }
+
+  private void downloadJxBrowser(Project project, String[] fileNames) {
+    // The FileDownloader API is used by other plugins - e.g.
+    // https://github.com/JetBrains/intellij-community/blob/b09f8151e0d189d70363266c3bb6edb5f6bfeca4/plugins/markdown/src/org/intellij/plugins/markdown/ui/preview/javafx/JavaFXInstallator.java#L48
+    final List<FileDownloader> fileDownloaders = new ArrayList<>();
+    final DownloadableFileService service = DownloadableFileService.getInstance();
+    for (String fileName : fileNames) {
+      final DownloadableFileDescription
+        description = service.createFileDescription(JxBrowserUtils.getDistributionLink(fileName), fileName);
+      fileDownloaders.add(service.createDownloader(Collections.singletonList(description), fileName));
+    }
+
+    final Task.Backgroundable task = new Task.Backgroundable(project, "Downloading jxbrowser") {
+      @Override
+      public void run(@NotNull ProgressIndicator indicator) {
+        String currentFileName = null;
+        try {
+          final List<File> files = new ArrayList<>();
+          for (int i = 0; i < fileDownloaders.size(); i++) {
+            final FileDownloader downloader = fileDownloaders.get(i);
+            currentFileName = fileNames[i];
+            final Pair<File, DownloadableFileDescription> download =
+              ContainerUtil.getFirstItem(downloader.download(new File(DOWNLOAD_PATH)));
+            final File file = download != null ? download.first : null;
+            if (file != null) {
+              files.add(file);
+              LOG.info(project.getName() + ": JxBrowser file downloaded: " + file.getAbsolutePath());
+            }
+          }
+
+          loadClasses(files);
+        }
+        catch (IOException e) {
+          LOG.info(project.getName() + ": JxBrowser file downloaded failed: " + currentFileName);
+          e.printStackTrace();
+          status.set(JxBrowserStatus.INSTALLATION_FAILED);
+        }
+      }
+    };
+    BackgroundableProcessIndicator processIndicator = new BackgroundableProcessIndicator(task);
+    processIndicator.setIndeterminate(false);
+    ProgressManager.getInstance().runProcessWithProgressAsynchronously(task, processIndicator);
+  }
+
+  private void loadClasses(List<File> files) {
+    final UrlClassLoader classLoader = (UrlClassLoader) this.getClass().getClassLoader();
+    try {
+      for (File file : files) {
+        final URL url = file.toURI().toURL();
+        classLoader.addURL(url);
+        LOG.info("Loaded JxBrowser file successfully: " + url.toString());
+      }
+
+      status.set(JxBrowserStatus.INSTALLED);
+    }
+    catch (Exception e) {
+      LOG.info("Failed to load JxBrowser files");
```

Comment:
what exception are you expecting? should we delete the cached browser files in this case?

Reply:
`toURL` can throw a malformed URL exception. I'm not sure how to handle this case. We could delete the files, but if this occurred because of a bug in our file naming, then it would be annoying for the user to have to keep downloading the files to a bad spot each time they opened their IDE.